sdk/core/Azure.Core/src/Shared/Utf8JsonReaderExtensions.cs

```diff
@@ -0,0 +1,166 @@
+            { typeof(float), (ref Utf8JsonReader reader) => reader.GetSingle() },
+            { typeof(decimal), (ref Utf8JsonReader reader) => reader.GetDecimal() },
+            { typeof(DateTimeOffset), (ref Utf8JsonReader reader) => reader.GetDateTimeOffset() },
+            { typeof(Guid), (ref Utf8JsonReader reader) => reader.GetGuid() },
+            { typeof(byte[]), (ref Utf8JsonReader reader) => reader.GetBytesFromBase64() },
+        };
+
+        private static readonly Dictionary<Type, Func<string, object>> _keyConverter = new Dictionary<Type, Func<string, object>>()
+        {
+            { typeof(int), str => Convert.ToInt32(str) },
+            { typeof(string), str => str },
+        };
+
+        private static string ConvertKeyToString(Type typeToConvert, string key) => key;
+
+        public static List<T> GetList<T>(this ref Utf8JsonReader reader, ModelSerializerOptions options)
+        {
+            List<T> list = new List<T>();
+
+            if (reader.TokenType == JsonTokenType.Null)
+                return null;
+
+            if (reader.TokenType != JsonTokenType.StartArray)
+                throw new InvalidOperationException("Expected StartArray token");
+
+            while (reader.Read())
+            {
+                if (reader.TokenType == JsonTokenType.EndArray)
+                    break;
+
+                list.Add(reader.GetObject<T>(options));
+            }
+
+            return list;
+        }
+
+        public static Dictionary<TKey, TValue> GetDictionary<TKey, TValue>(this ref Utf8JsonReader reader, ModelSerializerOptions options)
+            where TKey : notnull
+        {
+            if (!_keyConverter.TryGetValue(typeof(TKey), out var converter))
+                throw new InvalidOperationException($"Cannot use type {typeof(TKey).Name} as a key for a dictionary");
+
+            Dictionary<TKey, TValue> dictionary = new Dictionary<TKey, TValue>();
+            reader.Read();
+
+            if (reader.TokenType == JsonTokenType.Null)
+                return null;
+
+            if (reader.TokenType != JsonTokenType.StartObject)
+                throw new InvalidOperationException("Expected StartObject token");
+
+            while (reader.Read())
+            {
+                if (reader.TokenType == JsonTokenType.EndObject)
+                    break;
+
+                if (reader.TokenType != JsonTokenType.PropertyName)
+                    throw new InvalidOperationException("Expected PropertyName token");
+
+                var propertyName = reader.GetString();
+                reader.Read();
+                dictionary.Add((TKey)converter(propertyName), reader.GetObject<TValue>(options));
+            }
+
+            return dictionary;
+        }
+
+        public static T GetObject<T>(this ref Utf8JsonReader reader, ModelSerializerOptions options)
+        {
+            var typeToConvert = typeof(T);
+
+            if (_readerActions.TryGetValue(typeToConvert, out var readValue))
```

Comment:
This approach will cause boxing and unboxing for every value type.
If we want to use dictionary, we should create an interface of T and call its method, e.g.:

```cs
internal interface IKnownConverter<T>
{
    T GetKey(string key);
    T ReadValue(ref Utf8JsonReader reader);
}

internal class Int32KnownConverter : IKnownConverter<int>
{
    public int GetKey(string key) => Convert.ToInt32(str);
    public int ReadValue(ref Utf8JsonReader reader) => reader.GetInt32();
}

private static readonly Dictionary<Type, object> _knownConverters = new Dictionary<Type, object>
{
    [typeof(int)] = new Int32KnownConverter()
}

// Usage:
if (_knownConverters.TryGetValue(typeToConvert, out var converter))
{
    ((IKnownConverter<T>)converter).ReadValue(ref reader);
}
```

Reply:
https://github.com/Azure/azure-sdk-for-net/pull/37970